# Feedback from previous weeks and other and hints

1. Be careful about your environment remembering variables. Make sure your code works in a new _clean_ environment. In Colab: `Runtime`->`restart Runtime`, in Anaconda's Jupyter: `Kernel`->`Restart`.
2. Graphs without labels (or units when appropriate) are not worth any point.
3. Do put in sufficient explanatory comments in your code.
4. Functions are very important. Do look up the video on the Safari O'Reilly ressource if you are still not clear on them !

For this week you can use these imports at the start of your programs:

In [1]:
import numpy as np
import matplotlib.pyplot as plt

We will use a new module `uncertainties`, which is **not** standard in the colab environment. You will have to first run:

In [2]:
! pip install -q uncertainties

To install the module before you can import it:

In [3]:
import uncertainties as uc
import uncertainties.umath as um # for maths functions

# Introduction
In the practical classes PX2133/PX2233 and PX2338 (Obs tech), as well as your year 3/4 project, a lot of emphasis is placed on the determination and mathematical handling of errors.
The uncertainties module allows us to deal very easily with [error propagation](https://en.wikipedia.org/wiki/Propagation_of_uncertainty). For this sheet you should remind yourself about error bars in measurements and about propagation of uncertainties. Take an example from your lab handbook:

**Example 1**: If the length of a rectangle is $1.24\pm0.02 m$ and its breadth is $0.61\pm0.01 m$, what is its area and the error in the area? The following code snippet solves this problem in a few lines.

In [4]:
L = uc.ufloat(1.24, 0.02)
W = uc.ufloat(0.61, 0.01)
print ('Area is:', L*W, 'm^2') # Do remember to add the units when printing!

Area is: 0.756+/-0.017 m^2


**Note**: For the area itself, it's fairly straightforward:

In [5]:
1.24*0.61

0.7564

However, for the error bar on this number:

In [6]:
0.02*0.01

0.0002

does not work. Instead, the [error progation formula](https://en.wikipedia.org/wiki/Propagation_of_uncertainty#Example_formulae) gives:

In [7]:
np.abs(1.24*0.61)*np.sqrt((0.02/1.24)**2+(0.01/0.61)**2)

np.float64(0.01739540169125163)

So the area is $0.756\pm0.017 m^2$. `uncertainties` obviously saves a lot of work, even for such a simple case. You can also take a look at the web site uncertainties hosted at https://pythonhosted.org/uncertainties/user_guide.html. In particular, [this section](https://pythonhosted.org/uncertainties/user_guide.html#access-to-the-uncertainty-and-to-the-nominal-value) shows some of the properties of a `ufloat` you can access directly.

**Example 2**: A reference object is $10.0\pm0.0001 m$ long, and makes a viewing angle of $0.62\pm0.02 rad$. How far is it?

In [8]:
L = uc.ufloat(10.0, 0.0001)
theta = uc.ufloat(0.62,0.02)

Distance = (L/2)/um.tan(theta/2)

print ('Distance is:', Distance.nominal_value, 'm, with an error of:', Distance.std_dev)

Distance is: 15.609024890896208 m, with an error of: 0.537283338762715


Note the need to use "umath" functions (like `um.tan()` instead of `np.tan()`), and how to get the nominal value and the standard deviation of the uncertainties objects. To get nicer looking output, such as controlling the number of significant digits printed, you can use the information about formatting at https://docs.python.org/3/tutorial/inputoutput.html. In the exercises below you need to print the values to the screen. (Don’t forget units.)

# Exercises
This must be marked before you leave the lab. Mark weighting is in brackets.
**Save your work to GitHub after having run all cells with `Runtime` -> `Restart and run all`. And do not change the notebook's filename.** Do add comments to your code, you'll lose points if your code is hard to understand. Graphs without labels (or units when appropriate) are not worth any point.

## Exercise 0
[0] With some approximations, we have measured the mass of the following black-holes:
```
"35.6+/-3.9","30.6+/-3.7","63.1+/-3.2","23.2+/-9.8","13.6+/-4.5","35.7+/-6.8","13.7+/-6.0","7.7+/-2.4","20.5+/-4.0"
```
Compute for each (with error-bars) their lifetime due to Hawking radiation:
$$
t = \left(\frac{M}{M_{\odot}}\right)^3\,\times\,2.097\,\times\,10^{67} yr
$$

(this exercise is for demonstration purposes and won't be marked)

In [9]:
def lifetime(mass):
  '''
  Returns the lifetime in years of a black hole of mass m in solar masses
  '''
  return (mass**3)*2.097 * 1e67

In [10]:
strings = ["35.6+/-3.9","30.6+/-3.7","63.1+/-3.2","23.2+/-9.8","13.6+/-4.5","35.7+/-6.8","13.7+/-6.0","7.7+/-2.4","20.5+/-4.0"]

In [11]:
masses = [uc.ufloat_fromstr(s) for s in strings]

In [12]:
['{:10.1e}'.format(lifetime(bh_mass)) for bh_mass in masses]

['   9.5e+71+/-   3.1e+71',
 '   6.0e+71+/-   2.2e+71',
 '   5.3e+72+/-   0.8e+72',
 '   2.6e+71+/-   3.3e+71',
 '   5.3e+70+/-   5.2e+70',
 '   9.5e+71+/-   5.5e+71',
 '   5.4e+70+/-   7.1e+70',
 '   9.6e+69+/-   9.0e+69',
 '   1.8e+71+/-   1.1e+71']

## Exercise 1
[2] An object is measured to travel a distance $x = 5.1 \pm 0.4 m$ during a time of $t = 0.4 \pm 0.1 s$. What is the average velocity and the error in the average velocity?

In [13]:
! pip install -q uncertainties
import numpy as np
import matplotlib.pyplot as plt

import uncertainties as uc
import uncertainties.umath as um # for maths functions

# using the uc.ufloat function to calculate the average velocity and its error
x = uc.ufloat(5.1,0.4)
t = uc.ufloat(0.4, 0.1)

vel = x/t

print(f"The velocity is: {vel.nominal_value:.3f} m/s with an error of {vel.std_dev:.3f} m/s")
# nominal value is the first value in the uc.ufloat variable above for each component x and t
# std_dev is the standard deviation of the vel function, using both x and t components

The velocity is: 12.750 m/s with an error of 3.341 m/s


## Exercise 2
[2] An enterprising cow attempts to jump over the moon by jumping vertically into the air with initial speed $v_0=4.0\pm0.2 m/s$. After a time $t=0.60\pm0.06s$, the height of the cow is $h = v_0t-\frac{1}{2}g t^2 = 0.636 m$. What is the uncertainty in $h$? Take $g$ as exactly $9.81 ms^{-2}$.

In [14]:
! pip install -q uncertainties
import numpy as np
import matplotlib.pyplot as plt

import uncertainties as uc
import uncertainties.umath as um # for maths functions

# keeping aside the known variables
v_0 = uc.ufloat(4.0,0.2)
time = uc.ufloat(0.60,0.06)
g = 9.81
h_abs = 0.636

h_unc = v_0*t - 0.5*g*t**2

# printing out the uncertainty in height
print(f"The height is: {h_abs} m with an error of {h_unc.std_dev:.3f} m")

The height is: 0.636 m with an error of 0.080 m


## Exercise 3
[2] In an optics experiment the object distance $u$ is measured to be 20cm and the image distance $v$ is 10cm, both to an accuracy of 0.5cm. Find the focal length $f$ of the lens using the formula:

$$ \frac{1}{u}+\frac{1}{v}=\frac{1}{f}$$

In [15]:
! pip install -q uncertainties
import numpy as np
import matplotlib.pyplot as plt

import uncertainties as uc
import uncertainties.umath as um # for maths functions

# converting the distances to metres
u = uc.ufloat(0.2,0.005)
v = uc.ufloat(0.1,0.005)

# assuming the distances are in metres
f=1/((1/u)+(1/v))
print(f"The focal length, with u and v in m, is: {f.nominal_value:.3f} with an error of {f.std_dev:.3f}")

#converting back to centimeters
u_orig = uc.ufloat(20,0.5)
v_orig = uc.ufloat(10,0.5)
f_orig = 1/((1/u_orig)+(1/v_orig))

print(f"The focal length, with u and v in cm, is: {f_orig.nominal_value:.3f} with an error of {f_orig.std_dev:.3f}")

The focal length, with u and v in m, is: 0.067 with an error of 0.002
The focal length, with u and v in cm, is: 6.667 with an error of 0.229


## Exercise 4
[2] Two students each measure the refractive index of water. Jack measures a value of $1.33 \pm 0.03$ while Jill measures $1.28 \pm 0.02$. Are these values in agreement? *You do have to think a bit about this one...*

In [16]:
! pip install -q uncertainties
import numpy as np
import matplotlib.pyplot as plt

import uncertainties as uc
import uncertainties.umath as um # for maths functions

jack = uc.ufloat(1.33,0.03)
jill = uc.ufloat(1.28,0.02)

diff = jack.nominal_value - jill.nominal_value

u_combine = um.sqrt(jack.std_dev**2+jill.std_dev**2)

# I now compare the difference and the combined uncertainty
in_agreement = diff <= u_combine

# Printing the results
print(f"Difference between measurements: {diff:.3f}")
print(f"Combined uncertainty: {u_combine:.3f}")

if in_agreement:
    print("The measurements are in agreement.")
else:
    print("The measurements are not in agreement.")

Difference between measurements: 0.050
Combined uncertainty: 0.036
The measurements are not in agreement.


## Exercise 5
[2] The damped resonance frequency $\omega_{res}$ of an oscillating system is related to the (un-damped) natural angular frequency $\omega_0$ and the damping coefficient $\alpha$ by:

$$\omega_{res} = \sqrt{ \omega_0^2 - 2\alpha^2}$$

Find $f_0$ if the measured resonance frequency $f_{res}$ is $23.2\pm0.1 Hz$ and the measured damping coefficient is $19.5\pm0.5s^{-1}$.


In [17]:
! pip install -q uncertainties
import numpy as np
import matplotlib.pyplot as plt

import uncertainties as uc
import uncertainties.umath as um # for maths functions

#converting back to radians
omega_res = uc.ufloat(23.2*2*np.pi,0.1*2*np.pi)
alpha = uc.ufloat(19.5,0.5)

omega_0 = um.sqrt(omega_res**2+2*alpha**2)
f_0 = omega_0/2*np.pi

print(f"natural frequency is: {f_0.nominal_value:.3f} Hz with an error of {f_0.std_dev:.3f} Hz")

natural frequency is: 233.036 Hz with an error of 0.991 Hz


## Exercise 6
[2] Suppose you have the following equation from one of your lab experiments:

$$f=\frac{c}{2}\sqrt{\frac{n_x^2}{L_x^2}+\frac{n_y^2}{L_y^2}+\frac{n_z^2}{L_z^2}}$$

where $f$ is the resonant frequency of sound waves in a box of sides $L_x$, $L_y$ and $L_z$ in length and the $n_x$ etc. are integers. $L_x = 10.2\pm0.2m$, $L_y = 5.2\pm0.3m$ and $L_z = 20.0\pm0.1 m$, while $c = 331.3 + T * 0.606 \,m\,s^{-1}$ is the temperature-dependent speed of sound, and the temperature $T$ is $23 \pm 1^\circ C$.
Calculate $f$ and the error in $f$ for the following values of $(nx,ny,nz)= (1,1,1), (1,1,2)$ and $(2,1,1)$.


In [18]:
! pip install -q uncertainties
import numpy as np
import matplotlib.pyplot as plt

import uncertainties as uc
import uncertainties.umath as um # for maths functions

#writing the variables
'''
Dimensions of the box with uncertainties
'''
L_x = uc.ufloat(10.2,0.2)
L_y = uc.ufloat(5.2,0.3)
L_z = uc.ufloat(20.0,0.1)
'''
Temperature in degrees
'''
T = uc.ufloat(23,1) # in degrees

#speed of sound in a box, with T being defined above
c = 331.3 + T*0.606

''' Non integer values nx,ny,nz'''
#each n_value is an array of strings as I am asked to calculate 3 f values with uncertainties
n_1 = np.array([1,1,1])
n_2 = np.array([1,1,2])
n_3 = np.array([2,1,1])

f1 = c/2 *um.sqrt(n_1[0]**2/(L_x**2)+n_1[1]**2/L_y**2+n_1[2]**2/(L_z**2))
print("for (nx,ny,nz) = (1,1,1):")
print(f"res. frequency is: {f1.nominal_value:.3f} Hz with an error of {f1.std_dev:.3f} Hz \n")
f2 = c/2 *um.sqrt(n_2[0]**2/(L_x**2)+n_2[1]**2/L_y**2+n_2[2]**2/(L_z**2))
print("for (nx,ny,nz) = (1,1,2):")
print(f"res. frequency is: {f2.nominal_value:.3f} Hz with an error of {f2.std_dev:.3f} Hz \n")
f3 = c/2 *um.sqrt(n_3[0]**2/(L_x**2)+n_3[1]**2/L_y**2+n_3[2]**2/(L_z**2))
print("for (nx,ny,nz) = (2,1,1)")
print(f"res. frequency is: {f3.nominal_value:.3f} Hz with an error of {f3.std_dev:.3f} Hz")

for (nx,ny,nz) = (1,1,1):
res. frequency is: 38.247 Hz with an error of 1.670 Hz 

for (nx,ny,nz) = (1,1,2):
res. frequency is: 41.065 Hz with an error of 1.556 Hz 

for (nx,ny,nz) = (2,1,1)
res. frequency is: 48.188 Hz with an error of 1.402 Hz


## Exercise 7
[4] The reflection coefficient $R_\parallel$ for parallel plane-polarised light reflected from a surface is given by the equation:

$$ R_\parallel = \frac{\tan^2(\theta_i - \theta_t)}{\tan^2(\theta_i + \theta_t)} $$

Calculate the error in $R_\parallel$ given measurements $\theta_i = (78 \pm 1)^\circ$ and $\theta_t = (40 \pm 1)^\circ$.

In [29]:
! pip install -q uncertainties
import numpy as np
import matplotlib.pyplot as plt

import uncertainties as uc
import uncertainties.umath as um # for maths functions

t_i = uc.ufloat(78,1)
t_t = uc.ufloat(40,1)

R_parallel = (um.tan(t_i-t_t))**2/(um.tan(t_i+t_t))**2

print(f"reflection coefficient is: ",{R_parallel.nominal_value}, "with an uncertainty ",{R_parallel.std_dev})
print("to 4 decimal places, this is 0.0036 +/ 0.0651, which indicates high uncertainty")

reflection coefficient is:  {0.003572137026589645} with an uncertainty  {0.06506774439998199}
to 4 decimal places, this is 0.0036 +/ 0.0651, which indicates high uncertainty


## Exercise 8
[4] Calculate and print to the screen the fractional uncertainty, as a percentage to one
significant figure, of the fluid flow discharge coefficient $C_d$ from the equation

$$
C_d = \frac{\dot{m}\sqrt{1-\left(\frac{d}{D}\right)^4}}{Kd^2F\sqrt{\rho\Delta P}}
$$

where

\begin{align*}
    C_d &= \text{discharge coefficient}&& \text{(no units)} \\
    \dot{m} &= \text{mass flow rate}&& = 0.13 \pm 0.01kg\,s^{-1} \\
    d &= \text{orifice diameter}&& = 11\pm 1 mm \\
    D &= \text{pipe diameter}&& = 71 \pm 1 mm \\
    \rho &= \text{fluid density}&& =1.01\pm0.01g\,cm^{-3} \\
    \Delta P &= \text{differential pressure}&& =156 \pm 7 Pa \\
    K &= \text{a constant parameter}&& =\text{constant (no units)} \\
    F &= \text{thermal expansion factor}&& =\text{constant (no units)}
\end{align*}


In [33]:
! pip install -q uncertainties
import numpy as np
import matplotlib.pyplot as plt

import uncertainties as uc
import uncertainties.umath as um # for maths functions

#converting the values to SI units
mdot = uc.ufloat(0.13,0.01)
d = uc.ufloat(0.011,0.001)
D = uc.ufloat(0.071,0.001)
rho = uc.ufloat(1010,10)
P_delta = uc.ufloat(156,7)

#could not have time to do final part so this is how much I could do
C_d = (mdot*um.sqrt(1-(d/D)**4))/d**2*um.sqrt(rho*P_delta)

print(f"fluid flow discharge value is: ",{C_d.nominal_value}, "with an uncertainty ",{C_d.std_dev})

fluid flow discharge value is:  {426339.7459850153} with an uncertainty  {84777.32758565679}


## Exercise 9: Optional problem (not marked)
If you have time and want to try something interesting, do the following problem by plotting in 2D:
 - Draw an equilateral triangle with vertices and coordinates: vertex 1: $(p_1,q_1)$; vertex 2: $(p_2, q_2)$; vertex 3: $(p_3, q_3)$.
 - Place a dot at an arbitrary point $P = (x_0, y_0)$ within this triangle.
 - Find the next point by selecting randomly an integer $n = 1 , 2, $  or $3$ :
    1. If 1 , place a dot halfway between P and vertex 1.
    2. If 2 , place a dot halfway between P and vertex 2.
    3. If 3 , place a dot halfway between P and vertex 3.
 - Repeat the last two steps using the last dot as the new P.

Mathematically, the coordinates of successive points are given by the formulae

$$(x_{i+1},y_{i+1})=0.5[(x_i,y_i)+(p_n,q_n)]$$

and

$$n=int(1+3r_i),$$

where $r_i$ is a random number between 0 and 1 and where the $int()$ function outputs the closest integer smaller than or equal to the argument.

Try extending this to four vertices.